### PR TITLE
Gardena: Fixed channel update when using thing file config

### PR DESCRIPTION
**Pull request by gerrieg**
Signed-off-by: Gerhard Riegler <gerhard.riegler@gmail.com>

### CODE REVIEW

File: addons/binding/org.openhab.binding.gardena/src/main/java/org/openhab/binding/gardena/discovery/GardenaDeviceDiscoveryService.java [hunk 1]
```diff
@@ -158,8 +158,9 @@ public void deviceDiscovered(Device device) {
      * Removes the Gardena device.
      */
     public void deviceRemoved(Device device) {
-        ThingUID thingUID = UidUtils.generateThingUID(device, accountHandler.getThing());
```
Comment (martinvw) [inline addons/binding/org.openhab.binding.gardena/src/main/java/org/openhab/binding/gardena/discovery/GardenaDeviceDiscoveryService.java]: Do you still use `UidUtils.generateThingUID`

File: addons/binding/org.openhab.binding.gardena/src/main/java/org/openhab/binding/gardena/handler/GardenaAccountHandler.java [hunk 2]
```diff
@@ -178,21 +179,23 @@ public void handleCommand(ChannelUID channelUID, Command command) {
...
+                    gardenaThingHandler.updateProperties(device);
+                    for (Channel channel : gardenaThing.getChannels()) {
+                        gardenaThingHandler.updateChannel(channel.getUID());
+                    }
+                    gardenaThingHandler.updateStatus(device);
                 }
-                gardenaThingHandler.updateStatus(device);
+            } catch (GardenaException ex) {
+                logger.error("There is something wrong with your thing, please recreate the thing {}",
+                        gardenaThing.getUID(), ex);
```
Comment (martinvw) [inline addons/binding/org.openhab.binding.gardena/src/main/java/org/openhab/binding/gardena/handler/GardenaAccountHandler.java]: You might want to set if OFFLINE here?

File: addons/binding/org.openhab.binding.gardena/src/main/java/org/openhab/binding/gardena/handler/GardenaAccountHandler.java [hunk 3]
```diff
@@ -178,21 +179,23 @@ public void handleCommand(ChannelUID channelUID, Command command) {
...
-        try {
-            if (gardenaThing != null) {
-                GardenaThingHandler gardenaThingHandler = (GardenaThingHandler) gardenaThing.getHandler();
-                gardenaThingHandler.updateProperties(device);
-                for (Channel channel : gardenaThing.getChannels()) {
-                    gardenaThingHandler.updateChannel(channel.getUID());
+        for (ThingUID thingUID : UidUtils.getThingUIDs(device, getThing())) {
+            Thing gardenaThing = getThingByUID(thingUID);
+            try {
+                if (gardenaThing != null) {
```
Comment (martinvw) [inline addons/binding/org.openhab.binding.gardena/src/main/java/org/openhab/binding/gardena/handler/GardenaAccountHandler.java]: Can this happen, should it be handled in any way?

File: addons/binding/org.openhab.binding.gardena/src/main/java/org/openhab/binding/gardena/handler/GardenaAccountHandler.java [hunk 4]
```diff
@@ -178,21 +179,23 @@ public void handleCommand(ChannelUID channelUID, Command command) {
...
+                    for (Channel channel : gardenaThing.getChannels()) {
+                        gardenaThingHandler.updateChannel(channel.getUID());
+                    }
+                    gardenaThingHandler.updateStatus(device);
                 }
-                gardenaThingHandler.updateStatus(device);
+            } catch (GardenaException ex) {
+                logger.error("There is something wrong with your thing, please recreate the thing {}",
+                        gardenaThing.getUID(), ex);
+            } catch (AccountHandlerNotAvailableException ex) {
```
Comment (martinvw) [inline addons/binding/org.openhab.binding.gardena/src/main/java/org/openhab/binding/gardena/handler/GardenaAccountHandler.java]: Optionally name the `ex` `ignore` and maybe change the comment to include the reason why you ignored it?

File: addons/binding/org.openhab.binding.gardena/src/main/java/org/openhab/binding/gardena/util/UidUtils.java [hunk 5]
```diff
@@ -32,9 +36,31 @@ public static ThingUID generateThingUID(Device device, Bridge account) {
...
+     * Returns all ThingUIDs for a given device.
+     */
+    public static List<ThingUID> getThingUIDs(Device device, Bridge account) {
+        List<ThingUID> thingUIDs = new ArrayList<ThingUID>();
+        for (Thing thing : account.getThings()) {
+            String deviceId = thing.getConfiguration().as(GardenaDeviceConfig.class).deviceId;
+            if (deviceId == null) {
+                deviceId = thing.getUID().getId();
+            }
+            if (device.getId().equals(deviceId)) {
```
Comment (martinvw) [inline addons/binding/org.openhab.binding.gardena/src/main/java/org/openhab/binding/gardena/util/UidUtils.java]: You just guaranteed that the `deviceId` is not `null`, maybe switch the equals around?

**Comment by martinvw** [comment]
@gerrieg just a safety check, did you test it without the null check?

**Comment by gerrieg** [comment]
yes, you are right, it can't happen anymore

**Pull request merged**